### PR TITLE
Optimize switch on conditional access

### PR DESCRIPTION
**Pull request by alrz**
Fixes https://github.com/dotnet/roslyn/issues/21962

Generates the following:
```
  // Code size       28 (0x1c)
  .maxstack  1
  .locals init (int V_0)
  IL_0000:  ldarg.0
  IL_0001:  brfalse.s  IL_000e
  IL_0003:  ldarg.0
  IL_0004:  call       ""int string.Length.get""
  IL_0009:  stloc.0
  IL_000a:  ldloc.0
  IL_000b:  brfalse.s  IL_0015
  IL_000d:  ret
  IL_000e:  ldc.i4.0
  IL_000f:  call       ""void System.Console.Write(int)""
  IL_0014:  ret
  IL_0015:  ldc.i4.1
  IL_0016:  call       ""void System.Console.Write(int)""
  IL_001b:  ret
```
Instead of:
```
  // Code size 61 (0x3d)
  .maxstack 1
  .locals init (
      [0] valuetype [mscorlib]System.Nullable`1<int32>,
      [1] valuetype [mscorlib]System.Nullable`1<int32>,
      [2] int32
  )

  IL_0000: ldarg.0
  IL_0001: brtrue.s IL_000e
  IL_0003: ldloca.s 1
  IL_0005: initobj valuetype [mscorlib]System.Nullable`1<int32>
  IL_000b: ldloc.1
  IL_000c: br.s IL_0019
  IL_000e: ldarg.0
  IL_000f: call instance int32 [mscorlib]System.String::get_Length()
  IL_0014: newobj instance void valuetype [mscorlib]System.Nullable`1<int32>::.ctor(!0)
  IL_0019: stloc.0
  IL_001a: ldloca.s 0
  IL_001c: call instance bool valuetype [mscorlib]System.Nullable`1<int32>::get_HasValue()
  IL_0021: brfalse.s IL_002f
  IL_0023: ldloca.s 0
  IL_0025: call instance !0 valuetype [mscorlib]System.Nullable`1<int32>::GetValueOrDefault()
  IL_002a: stloc.2
  IL_002b: ldloc.2
  IL_002c: brfalse.s IL_0036
  IL_002e: ret
  IL_002f: ldc.i4.0
  IL_0030: call void [mscorlib]System.Console::Write(int32)
  IL_0035: ret
  IL_0036: ldc.i4.1
  IL_0037: call void [mscorlib]System.Console::Write(int32)
  IL_003c: ret
```
For this method:
```cs
    static void M(string str)
    {
        switch(str?.Length)
        {
            case null:
                Console.Write(0);
                break;
            case 0:
                Console.Write(1);
                break;
        }
    }
```

### CODE REVIEW

**Comment by alrz** [comment]
pattern-switch is undone, I suppose we should be handle it in `LowerDecisionTree(DecisionTree.ByType)` but currently we don't have access to the original expression. @gafter, can you confirm?

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 1]
```diff
@@ -1058,6 +1058,12 @@
     <Field Name="LabelExpressionOpt" Type="BoundLabel" Null="allow"/>
   </Node>
 
+  <!-- only used by codegen -->
+  <Node Name="BoundGotoExpression" Base="BoundExpression">
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: One of the major points for the expression/statement separation is that expressions do not interfere with control flow. The node breaks the important invariant and is added to the tree very early.

We need a different solution. Introducing goto expression is too big of a change, architecturally, to handle this issue.

Comment (alrz) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: This is a codegen-only node and is lowered similar to throw-expressions (emits a default to adjust invariants)

Also https://github.com/dotnet/csharplang/issues/176 would actually *rely* on the possibility that we can jump out of expressions.. I saw a comment from AlekseyTs worth to quote here (https://github.com/dotnet/roslyn/pull/24263#discussion_r162220656): "Short-circuiting binary operators imply a control flow." wouldn't that apply here?

**Comment by alrz** [comment]
Obsolete. we should reiterate on top of recursive patterns

**Pull request closed without merging**